### PR TITLE
dev/financial#143 Convert Paypal Pro to use Guzzle 

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Convert PaypalPro to use Guzzle in order to bring it in under unit testing.

https://lab.civicrm.org/dev/financial/-/issues/143

Before
----------------------------------------
No testing on Paypal Pro payment processing

After
----------------------------------------
Testing added

Technical Details
----------------------------------------

Note the following explanation is intended to help document the process. 

As with most refactors most of the work and most of the change is in the test

The way we bring payment processors under CI is by converting them to use guzzle for http requests (this also makes CURL preferred but optional a the the server level). Guzzle has support for using mockHandlers to simulate http requests and responses without actually making them - so once we have set up guzzle testing we can supply an expected response and not rely on an outside service. We can also test the contents of the outgoing test - this gives us quick additional cover (for example the tests where we pass PropertyBag to Authorize.net fail in our guzzle tests because we are checking the http request and the credit card number is missing).

The steps to convert and set up a test are basically:

1) create the test shell 
2) switch the code from curl to guzzle
3) switch to live credentials and comment out the mock handler. Use a breakpoint to capture the response from the provider
4) re-enable the handler, run again - it should fail because we are comparing 'placeholder' with the actual request now
-  use the failure from the actual request to put into the getExpectedSinglePaymentRequest() function

**1 create the test shell**
The initial test shell is pretty boilerplate. The guzzle helper class is available to extensions. The key when trying to capture the response is   [to comment out the line](https://github.com/civicrm/civicrm-core/pull/18350/files#diff-910a89da0e15185925b9cfd92cbba1d0R78)
```
$this->createMockHandler([$response]);
```

**2. switch the code from curl to guzzle**
[Actually pretty straight forward](https://github.com/civicrm/civicrm-core/pull/18350/files#diff-96a0ea19e8ac102ef100c35b3b3d1c31R1018-R1024). Some variable for different expectations by the processor - e.g content type of xml used in Authorize.net

At this stage out code looks pretty much like [the first](https://github.com/civicrm/civicrm-core/pull/18350/commits/15912a4decf275132d36ac7c8483805a49227e0e) of the 2 commits in this PR

**3 Capture the response**
Basically key processor credentials into the test so the request works in a test context, and then grab that string and plug it into the tests - these screenshots are from stepping through the PaypalImpl class in phpstorm

Screen Shot 2020-09-04 at 4 14 53 PM

Screen Shot 2020-09-04 at 4 16 53 PM

Alternatively you can call 
```
$this->getResponseBodies();
``` 
from your test once it has run through

Screen Shot 2020-09-04 at 5 18 46 PM

**Run the test & capture the requests that went out**
These will be in the test fail or $this->container in your test class. To view them from $this->container you should cast to a string or use the helper ``` $this->getRequestBodies()[0]```

### CODE REVIEW

File: CRM/Core/Payment/PayPalImpl.php [hunk 1]
```diff
@@ -45,6 +45,25 @@ public function __construct($mode, &$paymentProcessor) {
...
 
+  /**
+   * @var GuzzleHttp\Client
+   */
+  protected $guzzleClient;
+
+  /**
+   * @return \GuzzleHttp\Client
+   */
+  public function getGuzzleClient(): \GuzzleHttp\Client {
```
Comment (eileenmcnaughton) [inline CRM/Core/Payment/PayPalImpl.php:56]: Note how this means we can set our own guzzle client from the test

File: CRM/Core/Payment/PayPalImpl.php [hunk 2]
```diff
@@ -996,35 +1015,19 @@ public function invokeAPI($args) {
...
-    curl_setopt($ch, CURLOPT_URL, $url);
-    curl_setopt($ch, CURLOPT_VERBOSE, 0);
-
-    //turning off the server and peer verification(TrustManager Concept).
-    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, Civi::settings()->get('verifySSL'));
-    curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, Civi::settings()->get('verifySSL') ? 2 : 0);
-
-    curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
-    curl_setopt($ch, CURLOPT_POST, 1);
+     $response = (string) $this->getGuzzleClient()->post($url, [
```
Comment (eileenmcnaughton) [inline CRM/Core/Payment/PayPalImpl.php:1018]: This is the main conversion

File: CRM/Core/Payment/PayPalImpl.php [hunk 3]
```diff
@@ -996,35 +1015,19 @@ public function invokeAPI($args) {
...
-    curl_setopt($ch, CURLOPT_URL, $url);
-    curl_setopt($ch, CURLOPT_VERBOSE, 0);
-
-    //turning off the server and peer verification(TrustManager Concept).
-    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, Civi::settings()->get('verifySSL'));
-    curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, Civi::settings()->get('verifySSL') ? 2 : 0);
-
-    curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
-    curl_setopt($ch, CURLOPT_POST, 1);
+    $response = (string) $this->getGuzzleClient()->post($url, [
```
Comment (eileenmcnaughton) [inline CRM/Core/Payment/PayPalImpl.php:1018]: This is the actual change - for Authorize.net we used a slightly different variant to pass xml instead

**Comment by seamuslee001** [comment]
This looks fine to me

**Comment by eileenmcnaughton** [comment]
@kcristiano if you get to test the rc next month can you make a point of testing paypal pro? We made this change for Authorize.net with no issues so it should be fine

**Comment by kcristiano** [comment]
@eileenmcnaughton I'll add the PR to my RC site.   I think I have creds for paypal pro.  If not I'll DM you.
